service/tbc/mempool.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package tbc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/davecgh/go-spew/spew"
+)
+
+type mempool struct {
+	mtx sync.RWMutex
+
+	txs  map[chainhash.Hash][]byte // when nil, tx has not been downloaded
+	size int                       // total memory used by mempool
+}
+
+func (m *mempool) getDataConstruct(ctx context.Context) (*wire.MsgGetData, error) {
+	log.Tracef("getDataConstruct")
+	defer log.Tracef("getDataConstruct exit")
+
+	getData := wire.NewMsgGetData()
+
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+
+	for k, v := range m.txs {
+		if v != nil {
+			continue
+		}
+		if err := getData.AddInvVect(&wire.InvVect{
+			Type: wire.InvTypeTx,
+			Hash: k,
+		}); err != nil {
+			// only happens when asking max inventory
+			return nil, fmt.Errorf("construct get data: %w", err)
+		}
+	}
+	return getData, nil
+}
+
+func (m *mempool) txsInsert(ctx context.Context, msg *wire.MsgTx, raw []byte) error {
+	log.Tracef("txsInsert")
+	defer log.Tracef("txsInsert exit")
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if tx := m.txs[msg.TxHash()]; tx == nil {
```

Comment:
this will return `true` for both txs that are in the map and have a `nil` value as well as ones that are not in the map, is this what we want?

exemplified https://go.dev/play/p/RUqBqPkdr5R

Reply:
should we check `ok` instead?

https://go.dev/play/p/qqWF-vlgaSK

Reply:
the reason that I ask is because `invTxsInsert` sets keys to `nil`, so I assume there is a different between `nil` value and key not existing in the map.  I want to be sure this is safe with that assumption.

Reply:
Sharp observation but this is deliberate in order to deal with the the p2p protocol. You have to issue a inv, that returns *just* a hash. We use that condition to key off and issue a getdata. If tx == nil it means we have to still download it. This is awkward but correct.